Add cloud config support for TF_WORKSPACE (#30787)
hashicorp/terraform: website/docs/cli/cloud/settings.mdx

### Environment Variables

You can use environment variables to configure one or more `cloud` block attributes. This is helpful when you want to configure Terraform as part of a Continuous Integration (CI) pipeline. Terraform only reads these variables if the corresponding attribute is omitted from your configuration file. If you choose to configure the `cloud` block entirely through environment variables, you must still add an empty `cloud` block in your configuration file.

Use the following environment variables to configure the `cloud` block:

- `TF_ORGANIZATION` - The name of the organization. Serves as a fallback for `organization`
    in the cloud configuration. If both are specified, the configuration takes precedence. 

- `TF_HOSTNAME` - The hostname of a Terraform Enterprise installation. Serves as a fallback if `hostname` is not specified in the cloud configuration. If both are specified, the configuration takes precendence. 

> - `TF_WORKSPACE` - The workspace to select from the available set. The `cloud` block interprets this variable as the `name` value of the `workspaces` attribute. The `cloud` block only reads this variable if `workspaces` is not set in your configuration file. Terraform will continue to use this variable to select the workspace locally. Refer to [TF_WORKSPACE](https://www.terraform.io/cli/config/environment-variables#tf_workspace) for more details.

Reviewer: ```suggestion
- `TF_WORKSPACE` - The workspace to select from the available set. The `cloud` block interprets this variable as the `name` value of the `workspaces` attribute. The `cloud` block only reads this variable if the `workspaces` block is not included in your configuration file. When you supply `TF_WORKSPACE`, Terraform also uses it select the workspace locally. Refer to [TF_WORKSPACE](https://www.terraform.io/cli/config/environment-variables#tf_workspace) for more details.
```

> - `TF_WORKSPACE` - The workspace to select from. The `cloud` block interprets this variable as the `name` value of the `workspaces` attribute if the `workspaces` attrribute is not included in your configuration file. Furthermore, it validates that `TF_WORKSPACE` exists in the organization specified in your configuration or `TF_ORGANIZATION`. Alternatively, this variable can also be used to select a workspace if you specify tags in your `cloud` configuration, returning an error if `TF_WORKSPACE` is not included in the set of tags specified in your configuration. Refer to [TF_WORKSPACE](https://www.terraform.io/cli/config/environment-variables#tf_workspace) for more details.

Author: Docs sensei 🚀

> - `TF_WORKSPACE` - The workspace to select from. The `cloud` block interprets this variable as the `name` value of the `workspaces` attribute if the `workspaces` attrribute is not included in your configuration file. Furthermore, it validates that `TF_WORKSPACE` exists in the organization specified in your configuration or `TF_ORGANIZATION`. Alternatively, this variable can also be used to select a workspace if you specify tags in your `cloud` configuration, returning an error if `TF_WORKSPACE` is not included in the set of tags specified in your configuration. Refer to [TF_WORKSPACE](https://www.terraform.io/cli/config/environment-variables#tf_workspace) for more details.

Reviewer: lol thank you for taking my edits!! But - let me know about my question (you were so fast!!) :-)

### Environment Variables

> Alternatively, you can configure the `cloud` backend non-interactively using environment variables, such as when configuring Terraform in a CI pipeline. The `cloud` block can be configured with the following environment variables:

Reviewer: ```suggestion
You can use environment variables to configure one or more `cloud` block attributes. This is helpful when you want to configure Terraform as part of a Continuous Integration (CI) pipeline. Terraform only reads these variables if the corresponding attribute is omitted from your configuration file. If you choose to configure the `cloud` block entirely through environment variables, you must still add an empty `cloud` block in your configuration file.

Use the following environment variables to configure the `cloud` block:
```

> -> **Note:** These variables are only read if the corresponding attribute is omitted from your configuration. They behave as fallbacks rather than overriding your configuration. Furthermore, while all attributes can be omitted, an empty `cloud` block must still be present in your configuration.

Reviewer: i think we can just make this a part of the introductory context (refer to my edit above)
```suggestion
```

- `TF_ORGANIZATION` - The name of the organization. Serves as a fallback for `organization`
    in the cloud configuration. If both are specified, the configuration takes precedence. 

- `TF_HOSTNAME` - The hostname of a Terraform Enterprise installation. Serves as a fallback if `hostname` is not specified in the cloud configuration. If both are specified, the configuration takes precendence. 

> - `TF_WORKSPACE` - The workspace to select from the available set. The `cloud` block interprets this variable as the `name` value of the `workspaces` attribute. The cloud backend will only read this variable if `workspaces` is completely omitted from your configuration. It also does not alter the existing behavior of how Terraform interprets this variable. [Read more about TF_WORKSPACE](https://www.terraform.io/cli/config/environment-variables#tf_workspace)

Reviewer: A suggestion and then the last sentence confuses me a bit. I wonder if we should say something more like "Terraform will continue to use this variable to select the workspace locally. Refer to [Environment Variables](link) for details." 

I just think we should tell users briefly what the behavior IS - in case it affects them wanting to use this for the `cloud` block?

```suggestion
- `TF_WORKSPACE` - The workspace to select from the available set. The `cloud` block interprets this variable as the `name` value of the `workspaces` attribute. The `cloud` block only reads this variable if `workspaces` is not set in your configuration file. It also does not alter the existing behavior of how Terraform interprets this variable. [Read more about TF_WORKSPACE](https://www.terraform.io/cli/config/environment-variables#tf_workspace)
```

- `TF_HOSTNAME` - The hostname of a Terraform Enterprise installation. Serves as a fallback if `hostname` is not specified in the cloud configuration. If both are specified, the configuration takes precendence. 

> - `TF_WORKSPACE` - The workspace to select from the available set. The `cloud` block interprets this variable as the `name` value of the `workspaces` attribute allowing you to omit this attribute completely. It does not alter the existing behavior of how Terraform interprets this variable. [Read more about TF_WORKSPACE](https://www.terraform.io/cli/config/environment-variables#tf_workspace)

Reviewer: This is cool! Two questions for you:
1. Should we now mark "workspaces" argument as optional? Should organization also be marked similarly? And should we say that if those things aren't set, users WILL. be expected to supply that on the command line?
2. What happens if you have the arguments configured in the configuration and then try to supply these values also with command line variables? Do the command line variables override what's in the config?
3. The PR comments bring up something interesting - the idea that you could theoretically leave the cloud block empty in your config and interactively set it via these command line variables. I'm wondering whether we should add something to the docs telling users this and also recommending when it is and isn't appropriate to do that.

Reviewer 2: 1. I'm not sure about the first point. The argument is still technically required, we're just adding to the ways that data can be set.
2. The configuration file takes precedence over environment variables.
3. For a CI environment, I think the ideal case would be that the cloud configuration block be empty

Here's the [RFC](https://docs.google.com/document/d/1UHVz-_Xkom4RrcW4540Y_zKZpTs92oj0ZejIhHb-K3w) that covers the configuration environment variables. There is also an RFC that covers setting credentials with an environment variable, but that work has not yet started.

Author: Great questions!

On your first question, I have thought about this and my answer depends on which context we use to define the word `required`. The attribute (both `organization` and `workspaces`) is _required_ in the sense that the value must either exist in the configuration or as an environment variable. However it is not required to be explicitly defined in your configuration as it can be set as an environment variable. I think however that marking it as `optional` will indicate this value can be omitted completely, which is not true... the value must in fact still exist. 

On your second question, the configuration always takes precedence over the environment so you can't override your configuration with env variables. The environment variables behave more like a fallback option to make it easier for customers who want to configure Terraform non-interactively. From my understanding however, is that explicitly configuring Terraform in your configuration is still the recommended approach. 

Ditto on your third point! Always helpful to provide extra guidance with this feature.

Author: Perhaps I did not make in clear in the docs, but the cloud configuration will only look for `TF_WORKSPACE` if the `workspaces` block is completely omitted, so:

```hcl
terraform {
  cloud {
     workspaces {}
   }
}
```
is invalid even if the env variable is set and will yield an error.
